File: conformance/tests/p256-wallet/p256-wallet.runar.go
//go:build ignore

package contract

import "runar"

type P256Wallet struct {
	runar.SmartContract
	EcdsaPubKeyHash runar.Addr       `runar:"readonly"`
	P256PubKeyHash  runar.ByteString `runar:"readonly"`
}

func (c *P256Wallet) Spend(p256Sig runar.ByteString, p256PubKey runar.ByteString, sig runar.Sig, pubKey runar.PubKey) {
	runar.Assert(runar.Hash160(pubKey) == c.EcdsaPubKeyHash)
	runar.Assert(runar.CheckSig(sig, pubKey))
	runar.Assert(runar.Hash160(p256PubKey) == c.P256PubKeyHash)
	runar.Assert(runar.VerifyECDSAP256(sig, p256Sig, p256PubKey))
}

File: conformance/tests/p256-wallet/p256-wallet.runar.rs
use runar::prelude::*;

#[runar::contract]
struct P256Wallet {
    #[readonly]
    ecdsa_pub_key_hash: Addr,
    #[readonly]
    p256_pub_key_hash: ByteString,
}

#[runar::methods(P256Wallet)]
impl P256Wallet {
    #[public]
    fn spend(&self, p256_sig: &ByteString, p256_pub_key: &ByteString, sig: &Sig, pub_key: &PubKey) {
        assert!(hash160(pub_key) == self.ecdsa_pub_key_hash);
        assert!(check_sig(sig, pub_key));
        assert!(hash160(p256_pub_key) == self.p256_pub_key_hash);
        assert!(verify_ecdsa_p256(sig, p256_sig, p256_pub_key));
    }
}
